File: iss2021/it.unibo.resumableBoundaryWalker/backup/UsingConnTcp.java
package it.unibo.wenv;

import org.json.JSONObject;
import java.io.PrintWriter;
import java.net.Socket;

public class UsingConnTcp {
	private static String localHostName    = "localhost"; //"localhost"; 192.168.1.7
	private static String containerHostName= "wenv";
	private static int port                = 8999;
	private static String sep              =";";
 	protected static Socket clientSocket ;
	protected static PrintWriter outToServer;
	 
	public UsingConnTcp() {
	}
	protected void initClientConn( ) throws Exception {
		try { //attempt to connect for a local application
		 	clientSocket = new Socket(localHostName, port);
		 	outToServer  = new PrintWriter(clientSocket.getOutputStream());
		}catch (Exception e1) {
			//Perhaps we are in a container in which the virtualrobot is a service named containerHostName
			for( int i=1; i<=5; i++ ) {
				//we try several times, since the service wenv must be ready
				//See https://docs.docker.com/compose/startup-order/
				try {
						println("initClientConn attempt to connect in container ");
						clientSocket = new Socket(containerHostName, port);
						outToServer  = new PrintWriter(clientSocket.getOutputStream());
						break;
				}catch (Exception e) {
					println("initClientConn RETRY since ERROR: " + e.getMessage() );
					Thread.sleep(1000);
				}
			}//for
		}
	}

	public   void sendCmd(String msg) throws Exception {
		if( outToServer == null ) return;
		String jsonString = "{ 'type': '" + msg + "', 'arg': 800 }";
		JSONObject jsonObject = new JSONObject(jsonString);
		msg = sep+jsonObject.toString()+sep;
		System.out.println("sending msg=" + msg);
		outToServer.println(msg);
		outToServer.flush();
	}

	protected void println(String msg) {
		System.out.println(  msg);
	}

 	
	public   void mbotForward() {
 		try {  sendCmd("moveForward"); } catch (Exception e) {e.printStackTrace();}
	}
	public   void mbotBackward() {
		try { sendCmd("moveBackward"); } catch (Exception e) {e.printStackTrace();}
	}
	public   void mbotLeft() {
		try { sendCmd("turnLeft"); } catch (Exception e) {e.printStackTrace();}
	}
	public   void mbotRight(  ) {
		try { sendCmd("turnRight"); } catch (Exception e) {e.printStackTrace();}
	}
	public   void mbotStop() {
		try { sendCmd("alarm"); } catch (Exception e) {e.printStackTrace();}
	}

	public void doJob(){
		try {
			//Thread.sleep(5000);
			System.out.println("STARTING ... ");
			initClientConn();
			mbotForward();
			Thread.sleep(1000);
			mbotBackward();
			Thread.sleep(1000);

			mbotLeft() ;
			Thread.sleep(1000);
			mbotRight(  ) ;
			Thread.sleep(1000);
			/*
			mbotForward();
			Thread.sleep(1000);
			mbotForward();
			Thread.sleep(1000);
			 */
			mbotStop();
			//Thread.sleep(5000);	//avoids premature termination (in docker-compose)
			System.out.println("END");
		} catch (Exception e) {
			e.printStackTrace();
		}

	}
  	
//Just for testing
	public static void main(String[] args)   {
		new UsingConnTcp().doJob();
	}
	
 }

File: iss2021/it.unibo.resumableBoundaryWalker/backup/kotlin/it/unibo/temp/UsingConnTcp.kt
package it.unibo.temp

import org.json.JSONObject
import java.io.PrintWriter
import java.net.Socket

class UsingConnTcp {

    companion object {
        private const val localHostName = "localhost"
        public const val containerHostName = "wenv"
        private const val port = 8999
        private const val sep = ";"
        protected var clientSocket: Socket? = null
        protected var outToServer: PrintWriter? = null

        //Just for testing
        /*
        @JvmStatic
        fun main(args: Array<String>) {
            UsingConnTcp().doJob()
        }*/
    }

    @Throws(Exception::class)
    protected fun initClientConn() {
        try { //attempt to connect for a local application
            clientSocket = Socket(localHostName, port)
            outToServer = PrintWriter(clientSocket!!.getOutputStream())
        } catch (e1: Exception) {
            for (i in 1..10) { //See https://docs.docker.com/compose/startup-order/
                try {
                    println("initClientConn attempt $i to connect in container ")
                    clientSocket = Socket(containerHostName, port)
                    outToServer  = PrintWriter(clientSocket!!.getOutputStream())
                    break
                } catch (e: Exception) {
                    println("initClientConn RETRY since ERROR ${e.message } " )
                    Thread.sleep(1000)
                }
            }
        }
    }

    @Throws(Exception::class)
    fun sendCmd(msg1: String) {
        var msg = msg1
        if (outToServer == null) return
        val jsonString = "{ 'type': '$msg', 'arg': 800 }"
        val jsonObject = JSONObject(jsonString)
        msg = sep + jsonObject.toString() + sep
        println("sending msg=$msg")
        outToServer!!.println(msg)
        outToServer!!.flush()
    }

    fun mbotForward() {
        try {
            sendCmd("moveForward")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun mbotBackward() {
        try {
            sendCmd("moveBackward")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun mbotLeft() {
        try {
            sendCmd("turnLeft")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun mbotRight() {
        try {
            sendCmd("turnRight")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun mbotStop() {
        try {
            sendCmd("alarm")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun doJob() {
        try {
            //Thread.sleep(5000);
            println("STARTING ... ")
            initClientConn()
            mbotForward()
            Thread.sleep(1000)
            mbotBackward()
            Thread.sleep(1000)
            /*
			mbotLeft() ;
			Thread.sleep(1000);
			mbotForward();
			Thread.sleep(1000);
			mbotRight(  ) ;
			Thread.sleep(1000);
			mbotForward();
			Thread.sleep(1000);
			 */mbotStop()
            //Thread.sleep(5000);	//avoids premature termination (in docker-compose)
            println("END")
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}

fun main(args: Array<String>) {
    val sys = UsingConnTcp()
    sys.doJob()
}

